docker/scripts/cyber_start.sh

```diff
@@ -15,265 +15,509 @@
+    if [[ "${kernel}" != "Linux" ]]; then
+        warning "Running Apollo cyber container on ${kernel} is UNTESTED, exiting..."
+        exit 1
+    fi
+
+    local os_release="$(lsb_release -rs)"
+    case "${os_release}" in
+        14.04)
+            warning "[Deprecated] Support for Ubuntu 14.04 will be removed" \
+                    "in the near future. Please upgrade to ubuntu 18.04+."
+            ;;
+        16.04|18.04|20.04|*)
+            ## Question(storypku): Any special considerations here ?
+            if [[ "${HOST_ARCH}" == "${TARGET_ARCH}" ]]; then
+                volumes="${volumes} -v /dev:/dev"
+            fi
+            ;;
+    esac
+
+    volumes="${volumes} -v /etc/localtime:/etc/localtime:ro"
+    # volumes="${volumes} -v /usr/src:/usr/src"
+    if [[ "${HOST_ARCH}" == "${TARGET_ARCH}" ]]; then
+        volumes="${volumes} -v /dev/null:/dev/raw1394 \
+                            -v /media:/media \
+                            -v /tmp/.X11-unix:/tmp/.X11-unix:rw \
+                            -v /lib/modules:/lib/modules \
+                    "
+
+    fi
+    volumes="$(tr -s " " <<< "${volumes}")"
+    eval "${__retval}='${volumes}'"
+}
+
+function determine_display() {
+    local display
+    # read from env
+    if [[ -z "${DISPLAY}" ]]; then
         display=":0"
     else
         display="${DISPLAY}"
     fi
+    echo "${display}"
+}
 
-    setup_device
+function remove_existing_cyber_container() {
+    if docker ps -a --format '{{.Names}}' | grep -q "${CYBER_CONTAINER}"; then
+        info "Removing existing cyber container ${CYBER_CONTAINER}"
+        docker stop "${CYBER_CONTAINER}" >/dev/null
+        docker rm -v -f "${CYBER_CONTAINER}" >/dev/null
+    fi
+}
 
-    USER_ID=$(id -u)
-    GRP=$(id -g -n)
-    GRP_ID=$(id -g)
-    LOCAL_HOST=`hostname`
-    DOCKER_HOME="/home/$USER"
-    if [ "$USER" == "root" ];then
-        DOCKER_HOME="/root"
+function docker_pull_if_needed() {
+    local image="$1"
+    local use_local_image="$2"
+    if [[ ${use_local_image} -eq 1 ]]; then
+        info "Start cyber container based on local image: ${image}"
+    elif docker images -a --format '{{.Repository}}:{{.Tag}}' \
+        | grep -q "${image}"; then
+        info "Image ${image} found locally, will be used."
+        use_local_image=1
     fi
-    if [ ! -d "${CACHE_ROOT_DIR}" ]; then
-        mkdir "${CACHE_ROOT_DIR}"
+    if [[ ${use_local_image} -eq 1 ]]; then
+        return
     fi
 
-    info "Starting docker container \"${APOLLO_CYBER}\" ..."
+    # Note(storypku): use may-be-modified ${DOCKER_REPO}
+    image="${DOCKER_REPO}:${image##*:}"
+    echo "Start pulling docker image: ${image}"
+    if ! ${DOCKER_PULL_CMD} "${image}"; then
+        error "Failed to pull docker image: ${image}"
+        exit 1
+    fi
+}
 
-    DOCKER_CMD="nvidia-docker"
-    USE_GPU=1
-    if ! [ -x "$(command -v ${DOCKER_CMD})" ]; then
-        DOCKER_CMD="docker"
-        USE_GPU=0
+function check_multi_arch_support() {
```

Comment:
Experimental feature. Supportt for running container on a different architecture.